layers/+lang/factor/funcs.el

```diff
@@ -8,11 +8,116 @@
 ;; This file is not part of GNU Emacs.
 ;;
 ;;; License: GPLv3
+(require 'loadhist)
```

Comment:
I have three functions here that use something (non-autoloadable) from `loadhist`.  I could add a `(require 'loadhist)` at the beginning of each (redundant), or add the `(require 'loadhist)` in the `:config` part of the `fuel` package init function (semantically wrong place, also breaks stand-alone file integrity of `funcs.el`).

What is the preferred way?

Reply:
Well I have some more basic questions about this. Where does `loadhist` come from, is that part of fuel? If not how do you make sure it is already installed?

As to how and where to load it, can you say why the resource is not autoloadable? Basically you should be able to declare autoloads also from outside the package, we do this frequently with packages loaded by `use-package`. If this is really not possible then I prefer to move the require in the `:config` part of `init-fuel`. 

The reason for this is that the loading order of a layer is always funcs.el, config.el followed by package.el, so it is quite common to load `defuns` in funcs.el for which the prerequisites are supplied later by package.el. Also this follows the convention of `lazy loading everything` in spacemacs as `loadhist` is only loaded when `fuel` is finally loaded, given that `fuel` is properly autoloaded this only happens when someone really uses `fuel` the first time.

Reply:
> Well I have some more basic questions about this. Where does `loadhist` come from, is that part of fuel? If not how do you make sure it is already installed? 

`loadhist` is part of Emacs.

Only one of the used functions from that package is autoloadable, which is `unload-feature`.  As to why that is the case, I don't know. 

It has nothing to do with fuel. The functions could be called before fuel is used, because they are used to load/unload fuel, so it feels very strange to me to couple that to the `use-package` call.

> The reason for this is that the loading order of a layer is always funcs.el, config.el followed by package.el,

If I read `LAYERS.org` correctly, it states the order is `layers.el`, `package.el`, `funcs.el`, `config.el`.  Should not make a difference here, though.

Reply:
Ups you are right, my load order was wrong, thanks for correcting me :+1:.

Ok if this is with emacs it should still be allocated to some layer. As your layer is the only one using it it should be save to declare ownership for it by adding it to the package.el and mention it with an init function, in there you could call use-package this can then require something if necessary.

This would solve the load order issue and communicate clearly that this package is required for this layer. It also allows for later lazy loading of it when necessary.

Please have a look at the `restructuredtext` layer there the build in `rst-directives` is used in a similar way.

Reply:
I took a look at the `rst` layer.  I think in this case it actually makes sense to add autoloads instead (like e.g. the scala layer in `funcs.el`), since I am effectively only using some emacs library functions, and have no reason to provide an initializer for `loadhist`. The `:config` section of the `fuel` initializer would also not be a good choice, since this functionality explicitly deals with loading and unloading the package at run-time.